src/workerd/api/streams/standard.c++

```diff
@@ -3068,15 +3079,39 @@ jsg::Promise<void> WritableStreamJsController::pipeLoop(jsg::Lock& js) {
         source.release(js);
       }
       return rejectedMaybeHandledPromise<void>(js, reason, pipeThrough);
+    };
+
+    auto promise = write(js, result.value.map([&](jsg::Value& value) {
+      return value.getHandle(js.v8Isolate);
     }));
-  }), ioContext.addFunctor([this,ref=addRef()] (jsg::Lock& js, jsg::Value value) {
+
+    if (IoContext::hasCurrent()) {
+      auto& ioContext = IoContext::current();
+      return promise.then(js, ioContext.addFunctor(kj::mv(onSuccess)),
+                              ioContext.addFunctor(kj::mv(onFailure)));
+    } else {
+      return promise.then(js, kj::mv(onSuccess), kj::mv(onFailure));
+    }
+  };
+
+  auto onFailure = [this,ref=addRef()] (jsg::Lock& js, jsg::Value value) {
     // The read failed. We will handle the error at the start of the next iteration.
     return pipeLoop(js);
-  }));
+  };
+
+  auto promise = pipeLock.source.read(js);
+
+  if (IoContext::hasCurrent()) {
+    auto& ioContext = IoContext::current();
+    return promise.then(js, ioContext.addFunctor(kj::mv(onSuccess)),
+                            ioContext.addFunctor(kj::mv(onFailure)));
+  } else {
+    return promise.then(js, kj::mv(onSuccess), kj::mv(onFailure));
+  }
```

Comment:
This conditional continuation wrapping on promises could be worth factoring out.